src/orion/algo/robo/ablr/ablr_model.py

```diff
@@ -0,0 +1,589 @@
+        neg_mll, predictive_distribution_fn = self.predictive_mean_and_variance(
+            x_train, y_train
+        )
+
+        def predict_normalized(x_test: Tensor | np.ndarray) -> Normal:
+            x_test = torch.as_tensor(x_test, dtype=torch.float)
+            if self.normalize_inputs:
+                x_test = self._normalize_x(x_test)
+            y_pred_dist = predictive_distribution_fn(x_test)
+            y_pred_dist = self._unnormalize_y(y_pred_dist)
+            return y_pred_dist
+
+            # y_pred = y_pred_dist.rsample()
+            # return self.unnormalize_y(y_pred)
+
+        return neg_mll, predict_normalized
+
+    def _get_loss(self, x: np.ndarray, y: np.ndarray) -> Tensor:
+        """Gets the loss for the given input and target."""
+        return self(x, y)[0]
+
+    @overload
+    def train(self, mode: bool = True) -> ABLR:
+        """Set the model in train mode following (as a nn.Module subclass)."""
+        ...
+
+    # pylint: disable=arguments-differ
+    @overload
+    def train(self, X: np.ndarray, y: np.ndarray, do_optimize: bool = False) -> None:
+        """Train the algorithm (as a RoBO Model subclass)."""
+        ...
+
+    # pylint: disable=signature-differs
+    def train(
+        self,
+        *args,
+        **kwargs,
+    ):
+        if len(args) == 1 and isinstance(args[0], bool):
+            mode: bool = args[0]
+            return nn.Module.train(self, mode=mode)
+        if "mode" in kwargs:
+            mode: bool = kwargs["mode"]
+            return nn.Module.train(self, mode=mode)
+        bound_args = inspect.signature(self._train_robo).bind(*args, **kwargs)
+        bound_args.apply_defaults()
+        return self._train_robo(
+            X=bound_args.arguments["X"],
+            y=bound_args.arguments["y"],
+            do_optimize=bound_args.arguments["do_optimize"],
+        )
+
+    def _train_robo(
+        self, X: np.ndarray, y: np.ndarray, do_optimize: bool | None = None
+    ) -> None:
+        """Training method for the algorithm, as a RoBO Model subclass."""
+
+        if X is None or y is None:
+            raise RuntimeError(
+                "Invalid arguments. X and y can't be None, they should be numpy arrays."
+            )
+
+        # Save the dataset so we can use it to predict the mean and variance later.
+        # NOTE: This do_optimize isn't used, but is a parameter of the Robo base class.
+        self.X = torch.as_tensor(X, dtype=torch.float)
+        self.y = torch.as_tensor(y, dtype=torch.float).reshape([-1])
+        self.x_mean = self.X.mean(0)
+        self.x_var = self.X.var(0)
+        self.y_mean = self.y.mean(0)
+        self.y_var = self.y.var(0)
+
+        dataset = TensorDataset(self.X, self.y)
+
+        train_dataset = dataset
+        train_dataloader = DataLoader(
+            train_dataset, batch_size=self.hparams.batch_size, shuffle=True
+        )
+        # TODO: No validation dataset for now.
```

Comment:
What would the validation dataset used in this case? Where would there be a selection of HPs for the ABLR model?

Reply:
Sounds like a good idea. Would you want to turn this into an issue for later?

Reply:
Sure, sounds good.

Reply:
Done: https://github.com/Epistimio/orion.algo.robo/issues/11